std-bits/base/src/main/java/org/enso/base/FileLineReader.java

```diff
@@ -0,0 +1,402 @@
+
+    // ** Creates a preloaded stream from a byte array. */
+    public static ByteArrayOutputStreamWithContains fromByteArray(byte[] bytes) {
+      var stream = new ByteArrayOutputStreamWithContains(0);
+      stream.buf = bytes;
+      stream.count = bytes.length;
+      return stream;
+    }
+
+    // ** Based on https://www.geeksforgeeks.org/kmp-algorithm-for-pattern-searching/ */
+    public static int[] computeLongestPrefix(byte[] bytes) {
+      int[] longestPrefix = new int[bytes.length];
+
+      int i = 1;
+      int len = 0;
+      while (i < bytes.length) {
+        if (bytes[i] == bytes[len]) {
+          len++;
+          longestPrefix[i++] = len;
+        } else if (len == 0) {
+          longestPrefix[i++] = 0;
+        } else {
+          len = longestPrefix[len - 1];
+        }
+      }
+
+      return longestPrefix;
+    }
+
+    public boolean contains(byte[] bytes, int[] longestPrefix) {
+      if (bytes.length > count) {
+        return false;
+      }
+
+      int i = 0;
+      int j = 0;
+      while ((count - i) >= (bytes.length - j)) {
+        if (buf[i] == bytes[j]) {
+          i++;
+          j++;
+        }
+
+        if (j == bytes.length) {
+          return true;
+        }
+
+        if (i < count && buf[i] != bytes[j]) {
+          if (j != 0) {
+            j = longestPrefix[j - 1];
+          } else {
+            i++;
+          }
+        }
+      }
+
+      return false;
+    }
+  }
+
+  private static class CancellationToken {
+    public boolean isCancelled = false;
+
+    public void cancel() {
+      isCancelled = true;
+    }
+  }
+
+  private static final Logger LOGGER = Logger.getLogger("enso-file-line-reader");
+
+  // ** Amount of data to read at a time for a single line (4KB). */
+  private static final int LINE_BUFFER = 4 * 1024;
+
+  // ** Amount of data to read at a time (4MB). */
+  private static final int BUFFER_SIZE = 4 * 1024 * 1024;
+
+  private static boolean moreToRead(int c, MappedByteBuffer buffer) {
+    return switch (c) {
+      case '\n', -1 -> false;
+      case '\r' -> {
+        c = buffer.hasRemaining() ? buffer.get() : '\n';
+        if (c != '\n') {
+          buffer.position(buffer.position() - 1);
+        }
+        yield false;
+      }
+      default -> true;
+    };
+  }
+
+  private static int readByte(MappedByteBuffer buffer) {
+    return buffer.hasRemaining() ? buffer.get() : -1;
+  }
+
+  // ** Reads a line into an OutputStream.
+  //    Returns true if the end of the line was found, false if the buffer finished. */
+  private static boolean readLine(MappedByteBuffer buffer, ByteArrayOutputStream result) {
+    int c = readByte(buffer);
+    while (moreToRead(c, buffer)) {
+      result.write(c);
+      c = readByte(buffer);
+      Context.getCurrent().safepoint();
+    }
+    return c != -1 && (c != '\r' || buffer.hasRemaining());
+  }
+
+  // ** Scans forward one line.
+  //    Returns true if the end of the line was found, false if the buffer finished. */
```

Comment:
Can we turn these into block comments as is the practice in most of our codebase?